File: src/cli/compare-true-cost.ts
#!/usr/bin/env npx ts-node
/**
 * True Cost Comparison CLI
 *
 * Compares offers by total cost including package price + baggage surcharge + transport cost.
 * Uses Turso for offers, hotel areas, and transport routes. Airline metadata
 * remains in the skill reference ota-knowledge.json.
 *
 * Usage:
 *   npx ts-node src/cli/compare-true-cost.ts --region kansai --date 2026-02-24 --pax 2
 *   npx ts-node src/cli/compare-true-cost.ts --region kansai --date 2026-02-24 --itinerary "kyoto:1,osaka:2"
 */

import * as fs from 'fs';
import * as path from 'path';
import { loadHotelAreasFromTurso, loadTransportRoutesFromTurso, queryOffers } from '../services/turso-service';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface TransportRoute {
  time_min: number;
  cost_jpy: number;
  method: string;
}

interface TransportRoutes {
  [region: string]: {
    routes: Record<string, TransportRoute>;
    hubs: Record<string, { type: string; area: string }>;
  };
}

interface AirlineInfo {
  code: string;
  type: 'LCC' | 'FSC';
  hand_baggage_kg: number;
  checked_bag_included: boolean;
  checked_bag_cost_twd?: number;
  checked_bag_kg?: number;
}

interface TrueCostOffer {
  file: string;
  source_name: string;
  price_per_person: number;
  currency: string;
  airline: string;
  hotel: string;
  hotel_area_type: string;
  baggage_cost: number;
  transport_cost: number;
  transport_time_min: number;
  true_total: number;
  breakdown: string;
}

// ---------------------------------------------------------------------------
// Data loaders
// ---------------------------------------------------------------------------

function loadJson<T>(filePath: string): T | null {
  try {
    return JSON.parse(fs.readFileSync(filePath, 'utf-8'));
  } catch { return null; }
}

function loadTransportRoutes(): TransportRoutes {
  throw new Error('loadTransportRoutes must use Turso via loadTransportRoutesFromTurso()');
}

function loadHotelAreas(): Record<string, Record<string, string[]>> {
  throw new Error('loadHotelAreas must use Turso via loadHotelAreasFromTurso()');
}

function loadAirlines(): Record<string, AirlineInfo> {
  const knowledge = loadJson<any>(
    path.join(process.cwd(), 'src', 'skills', 'travel-shared', 'references', 'ota-knowledge.json')
  );
  return knowledge?.airlines || {};
}

// ---------------------------------------------------------------------------
// Hotel area detection (mirrors Python version)
// ---------------------------------------------------------------------------

function detectHotelArea(hotelName: string, region: string, hotelAreas: Record<string, Record<string, string[]>>): string {
  const areas = hotelAreas[region] || {};
  for (const [areaType, keywords] of Object.entries(areas)) {
    for (const kw of keywords) {
      if (hotelName.includes(kw)) return areaType;
    }
  }
  return 'unknown';
}

// Map hotel area type to nearest transport hub
function areaTypeToHub(areaType: string, region: string, routes: TransportRoutes): string {
  const hubs = routes[region]?.hubs || {};
  // Find first hub matching this area type
  for (const [hubId, info] of Object.entries(hubs)) {
    if (info.type === areaType) return hubId;
  }
  // Default to first central hub
  for (const [hubId, info] of Object.entries(hubs)) {
    if (info.type === 'central') return hubId;
  }
  return '';
}

// ---------------------------------------------------------------------------
// Cost calculators
// ---------------------------------------------------------------------------

function lookupAirline(airlineName: string, airlines: Record<string, AirlineInfo>): AirlineInfo | null {
  const lower = airlineName.toLowerCase();
  // Direct key match
  for (const [key, info] of Object.entries(airlines)) {
    if (lower.includes(key.replace(/_/g, ' ')) || lower.includes(info.code.toLowerCase())) {
      return info;
    }
  }
  // Common name mappings
  const nameMap: Record<string, string> = {
    '樂桃': 'peach', 'peach': 'peach',
    '虎航': 'tigerair', 'tigerair': 'tigerair', '台灣虎航': 'tigerair',
    '捷星': 'jetstar', 'jetstar': 'jetstar',
    '長榮': 'eva', 'eva': 'eva', 'eva air': 'eva',
    '華航': 'china_airlines', '中華航空': 'china_airlines', 'china airlines': 'china_airlines',
    '星宇': 'starlux', 'starlux': 'starlux',
    'thai lion': 'thai_lion',
    'thai vietjet': 'thai_vietjet', '泰越捷': 'thai_vietjet', 'vietjet': 'thai_vietjet',
  };
  for (const [name, key] of Object.entries(nameMap)) {
    if (lower.includes(name)) return airlines[key] || null;
  }
  return null;
}

function calcBaggageCost(
  airline: string,
  packageType: string,
  baggageIncluded: boolean | null,
  pax: number,
  airlines: Record<string, AirlineInfo>,
): number {
  // Explicit baggage status overrides package type
  if (baggageIncluded === true) return 0;
  if (baggageIncluded === false) {
    // EzTravel FIT doesn't include baggage - use airline lookup
    const info = lookupAirline(airline, airlines);
    if (!info || info.checked_bag_included) return 0;
    const costPerSegment = info.checked_bag_cost_twd || 600;
    return 2 * pax * costPerSegment;
  }

  // Null/unknown: FIT packages typically include baggage
  if (packageType === 'fit') return 0;

  const info = lookupAirline(airline, airlines);
  if (!info) return 0;
  if (info.checked_bag_included) return 0;

  // LCC without included baggage: 2 directions × pax × cost
  const costPerSegment = info.checked_bag_cost_twd || 600;
  return 2 * pax * costPerSegment;
}

function lookupRoute(from: string, to: string, regionRoutes: Record<string, TransportRoute>): TransportRoute | null {
  const key1 = `${from}-${to}`;
  const key2 = `${to}-${from}`;
  return regionRoutes[key1] || regionRoutes[key2] || null;
}

function calcTransportCost(
  hotelHub: string,
  itinerary: Map<string, number>,
  region: string,
  pax: number,
  routes: TransportRoutes,
  jpy2twd: number,
): { cost: number; time_min: number } {
  const regionData = routes[region];
  if (!regionData || !hotelHub) return { cost: 0, time_min: 0 };

  let totalJpy = 0;
  let totalMin = 0;

  // Airport transfer: 2 trips (arrival + departure)
  const airportHub = Object.entries(regionData.hubs).find(([, v]) => v.type === 'airport')?.[0];
  if (airportHub) {
    const airportRoute = lookupRoute(airportHub, hotelHub, regionData.routes);
    if (airportRoute) {
      totalJpy += airportRoute.cost_jpy * pax * 2;
      totalMin += airportRoute.time_min * 2;
    }
  }

  // Daily itinerary transport
  for (const [dest, days] of itinerary) {
    const route = lookupRoute(hotelHub, dest, regionData.routes);
    if (route) {
      // Round trip per day
      totalJpy += route.cost_jpy * pax * 2 * days;
      totalMin += route.time_min * 2 * days;
    }
  }

  return { cost: Math.round(totalJpy * jpy2twd), time_min: totalMin };
}

// ---------------------------------------------------------------------------
// Offer scanner (simplified from travel-update.ts)
// ---------------------------------------------------------------------------

interface RawOffer {
  file: string;
  source_id: string;
  price_per_person: number;
  currency: string;
  airline: string;
  hotel: string;
  package_type: string;
  baggage_included: boolean | null;
}

async function scanOffers(region: string, filterDate: string | undefined, pax: number): Promise<RawOffer[]> {
  const rows = await queryOffers({
    region,
    ...(filterDate ? { start: filterDate, end: filterDate } : {}),
    type: 'package',
    limit: 500,
  });

  const offers: RawOffer[] = rows
    .filter((row) => row.price_per_person !== null)
    .map((row) => ({
      file: row.id,
      source_id: row.source_id,
      price_per_person: row.price_per_person!,
      currency: row.currency || 'TWD',
      airline: row.airline || '',
      hotel: row.hotel_name || row.name || '',
      package_type: 'fit',
      baggage_included: null,
    }));
  offers.sort((a, b) => a.price_per_person - b.price_per_person);
  return offers;
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

function parseItinerary(spec: string): Map<string, number> {
  const map = new Map<string, number>();
  for (const part of spec.split(',')) {
    const [dest, days] = part.trim().split(':');
    if (dest && days) map.set(dest.trim(), parseInt(days, 10));
  }
  return map;
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const opt = (name: string) => {
    const i = args.indexOf(name);
    return i >= 0 && i + 1 < args.length ? args[i + 1] : undefined;
  };

  const region = opt('--region');
  if (!region) {
    console.error('Usage: compare-true-cost --region <name> [--date YYYY-MM-DD] [--pax N] [--itinerary "kyoto:1,osaka:2"] [--jpy-rate N]');
    process.exit(1);
  }

  const filterDate = opt('--date');
  const pax = parseInt(opt('--pax') || '2', 10);
  const itinerarySpec = opt('--itinerary');
  const jpyRate = parseFloat(opt('--jpy-rate') || '0.22'); // JPY → TWD

  // Load reference data
  const routes = await loadTransportRoutesFromTurso() as TransportRoutes;
  const hotelAreas = await loadHotelAreasFromTurso();
  const airlines = loadAirlines();

  // Scan offers
  const rawOffers = await scanOffers(region, filterDate, pax);
  if (rawOffers.length === 0) {
    console.log(`No offers found for region "${region}".`);
    process.exit(1);
  }

  // Parse itinerary destinations
  const itinerary = itinerarySpec ? parseItinerary(itinerarySpec) : new Map<string, number>();

  // Calculate true cost for each offer
  const results: TrueCostOffer[] = [];

  for (const o of rawOffers) {
    const areaType = o.hotel
      ? detectHotelArea(o.hotel, region, hotelAreas)
      : 'unknown';
    const hotelHub = areaTypeToHub(areaType, region, routes);

    const baggageCost = calcBaggageCost(o.airline, o.package_type, o.baggage_included, pax, airlines);

    const transport = calcTransportCost(hotelHub, itinerary, region, pax, routes, jpyRate);

    const trueTotalPerPerson = o.price_per_person + Math.round((baggageCost + transport.cost) / pax);

    results.push({
      file: o.file,
      source_name: o.source_id,
      price_per_person: o.price_per_person,
      currency: o.currency,
      airline: o.airline,
      hotel: o.hotel,
      hotel_area_type: areaType,
      baggage_cost: baggageCost,
      transport_cost: transport.cost,
      transport_time_min: transport.time_min,
      true_total: trueTotalPerPerson,
      breakdown: `pkg:${o.price_per_person} bag:${baggageCost} xport:${transport.cost}`,
    });
  }

  results.sort((a, b) => a.true_total - b.true_total);

  // Print results
  const bestTime = Math.min(...results.map(r => r.transport_time_min));

  console.log(`\n╔════════════════════════════════════════════════════════════════════════════════════╗`);
  console.log(`║  TRUE COST COMPARISON: ${region.toUpperCase().padEnd(57)}║`);
  console.log(`╚════════════════════════════════════════════════════════════════════════════════════╝`);
  console.log(`  Pax: ${pax} | Date: ${filterDate || '(best)'} | JPY rate: ${jpyRate}`);
  if (itinerarySpec) console.log(`  Itinerary: ${itinerarySpec}`);
  console.log('');

  console.log('┌──────────┬──────────┬─────────┬───────────┬────────────┬──────────┬──────────────────────┐');
  console.log('│ Source   │ Package  │ Baggage │ Transport │ TRUE/person│ Time     │ Hotel                │');
  console.log('├──────────┼──────────┼─────────┼───────────┼────────────┼──────────┼──────────────────────┤');

  for (const r of results) {
    const src = r.source_name.slice(0, 8).padEnd(8);
    const pkg = r.price_per_person.toLocaleString().padStart(8);
    const bag = r.baggage_cost > 0 ? r.baggage_cost.toLocaleString().padStart(7) : '      0';
    const xport = r.transport_cost > 0 ? r.transport_cost.toLocaleString().padStart(9) : '        0';
    const total = r.true_total.toLocaleString().padStart(10);
    const timeDiff = r.transport_time_min - bestTime;
    const timeStr = timeDiff === 0 ? 'optimal ' : `+${timeDiff}min`.padEnd(8);
    const hotel = (r.hotel || '-').slice(0, 20).padEnd(20);

    console.log(`│ ${src} │ ${pkg} │ ${bag} │ ${xport} │ ${total} │ ${timeStr} │ ${hotel} │`);
  }

  console.log('└──────────┴──────────┴─────────┴───────────┴────────────┴──────────┴──────────────────────┘');

  // Recommendation
  if (results.length > 0) {
    const best = results[0];
    console.log(`\n💡 Best true value: ${best.source_name} at ${best.currency} ${best.true_total.toLocaleString()}/person`);
    console.log(`   ${best.breakdown}`);
    if (best.hotel) console.log(`   Hotel: ${best.hotel} (${best.hotel_area_type})`);
    if (best.airline) console.log(`   Airline: ${best.airline}`);

    // Warn if cheapest package != cheapest true cost
    const cheapestPkg = [...results].sort((a, b) => a.price_per_person - b.price_per_person)[0];
    if (cheapestPkg.file !== best.file) {
      console.log(`\n⚠️  Cheapest package (${cheapestPkg.source_name} ${cheapestPkg.price_per_person.toLocaleString()}) is NOT the best true value!`);
      console.log(`   Hidden costs: baggage ${cheapestPkg.baggage_cost.toLocaleString()} + transport ${cheapestPkg.transport_cost.toLocaleString()}`);
    }
  }

  console.log('');
}

main().catch((err) => {
  console.error(err instanceof Error ? err.message : String(err));
  process.exit(1);
});

File: rust/crates/travel-cli/src/compare_true_cost.rs
// `travel compare true-cost` — compare package offers by TOTAL cost (package +
// baggage surcharge + transport). Read-only, plain-text box tables, byte-parity
// with src/cli/compare-true-cost.ts. Reads offers + transport_routes/hubs +
// airlines + hotel_area_keywords (the de-JSON'd child table) from Turso.

use crate::db;
use std::collections::BTreeMap;

const DEFAULT_PAX: i64 = 2;
const DEFAULT_JPY_TWD: f64 = 0.22;

pub struct TrueCostArgs {
    pub region: String,
    pub date: Option<String>,
    pub pax: i64,
    pub itinerary: Option<String>,
    pub jpy_rate: f64,
}

impl TrueCostArgs {
    pub fn parse(args: &[String]) -> Result<Self, String> {
        let mut region: Option<String> = None;
        let mut date = None;
        let mut pax = DEFAULT_PAX;
        let mut itinerary = None;
        let mut jpy_rate = DEFAULT_JPY_TWD;
        let mut i = 0;
        while i < args.len() {
            let key = args[i].as_str();
            let val = || args.get(i + 1).cloned().ok_or_else(|| format!("{key} requires a value"));
            match key {
                "--region" => region = Some(val()?),
                "--date" => date = Some(val()?),
                "--pax" => pax = val()?.parse().map_err(|_| "--pax must be an integer".to_string())?,
                "--itinerary" => itinerary = Some(val()?),
                "--jpy-rate" => jpy_rate = val()?.parse().map_err(|_| "--jpy-rate must be a number".to_string())?,
                other => return Err(format!("unknown flag for compare true-cost: {other}")),
            }
            i += 2;
        }
        let region = region.ok_or_else(|| {
            "Usage: compare-true-cost --region <name> [--date YYYY-MM-DD] [--pax N] [--itinerary \"kyoto:1,osaka:2\"] [--jpy-rate N]".to_string()
        })?;
        Ok(TrueCostArgs { region, date, pax, itinerary, jpy_rate })
    }
}

fn sql_quote(v: &str) -> String {
    v.replace('\'', "''")
}

// JS Math.round: half rounds toward +Infinity (not away-from-zero).
fn js_round(x: f64) -> i64 {
    (x + 0.5).floor() as i64
}

// locale string (thousands separators) for integers — matches Number.toLocaleString.
fn locale(n: i64) -> String {
    let neg = n < 0;
    let digits = n.abs().to_string();
    let bytes = digits.as_bytes();
    let mut out = String::new();
    for (i, b) in bytes.iter().enumerate() {
        if i > 0 && (bytes.len() - i).is_multiple_of(3) {
            out.push(',');
        }
        out.push(*b as char);
    }
    if neg { format!("-{out}") } else { out }
}

// pad helpers operating on char count (JS string length), right/left.
fn pad_end(s: &str, n: usize) -> String {
    let len = s.chars().count();
    if len >= n { s.to_string() } else { format!("{}{}", s, " ".repeat(n - len)) }
}
fn pad_start(s: &str, n: usize) -> String {
    let len = s.chars().count();
    if len >= n { s.to_string() } else { format!("{}{}", " ".repeat(n - len), s) }
}
fn truncate(s: &str, n: usize) -> String {
    s.chars().take(n).collect()
}

struct TransportRoute {
    time_min: i64,
    cost_jpy: i64,
}

struct RegionData {
    routes: BTreeMap<String, TransportRoute>,
    hubs: Vec<(String, String, String)>, // (hub_id, hub_type, area) — insertion order
}

pub async fn run(opts: &TrueCostArgs) -> Result<(), String> {
    let conn = db::connect_read().await?;

    // NOTE: scanOffers hardcodes package_type='fit'/baggage_included=null, so the
    // TS calcBaggageCost always returns 0 here and never consults the airlines
    // table. We therefore skip loading airlines entirely (parity-equivalent).

    // --- transport routes + hubs ---
    let mut regions: BTreeMap<String, RegionData> = BTreeMap::new();
    {
        let mut rows = conn.query("SELECT region, route_key, time_min, cost_jpy FROM transport_routes ORDER BY region, route_key", ()).await.map_err(|e| format!("routes query: {e}"))?;
        while let Some(r) = rows.next().await.map_err(|e| format!("route row: {e}"))? {
            let region: String = r.get(0).unwrap_or_default();
            let key: String = r.get(1).unwrap_or_default();
            let entry = regions.entry(region).or_insert_with(|| RegionData { routes: BTreeMap::new(), hubs: Vec::new() });
            entry.routes.insert(key, TransportRoute { time_min: r.get(2).unwrap_or(0), cost_jpy: r.get(3).unwrap_or(0) });
        }
        let mut rows = conn.query("SELECT region, hub_id, hub_type, area FROM transport_hubs ORDER BY region, hub_id", ()).await.map_err(|e| format!("hubs query: {e}"))?;
        while let Some(r) = rows.next().await.map_err(|e| format!("hub row: {e}"))? {
            let region: String = r.get(0).unwrap_or_default();
            let entry = regions.entry(region).or_insert_with(|| RegionData { routes: BTreeMap::new(), hubs: Vec::new() });
            entry.hubs.push((r.get(1).unwrap_or_default(), r.get(2).unwrap_or_default(), r.get(3).unwrap_or_default()));
        }
    }

    // --- hotel area keywords (de-JSON'd child table), grouped region → areaType → [kw] ---
    // BTreeMap keeps areaType order stable; within an area, sort_order preserved by ORDER BY.
    let mut hotel_areas: BTreeMap<String, BTreeMap<String, Vec<String>>> = BTreeMap::new();
    {
        let mut rows = conn.query("SELECT region, area_type, keyword FROM hotel_area_keywords ORDER BY region, area_type, sort_order", ()).await.map_err(|e| format!("hotel_area_keywords query: {e}"))?;
        while let Some(r) = rows.next().await.map_err(|e| format!("hotel area row: {e}"))? {
            let region: String = r.get(0).unwrap_or_default();
            let area_type: String = r.get(1).unwrap_or_default();
            let kw: String = r.get(2).unwrap_or_default();
            hotel_areas.entry(region).or_default().entry(area_type).or_default().push(kw);
        }
    }

    // --- offers (package, region, optional exact date) ---
    let mut conds = vec![format!("region = '{}'", sql_quote(&opts.region)), "type = 'package'".to_string()];
    if let Some(d) = &opts.date {
        conds.push(format!("departure_date >= '{}'", sql_quote(d)));
        conds.push(format!("departure_date <= '{}'", sql_quote(d)));
    }
    let sql = format!(
        "SELECT id, source_id, price_per_person, currency, airline, hotel_name, name FROM offers \
         WHERE {} ORDER BY scraped_at DESC, price_per_person ASC LIMIT 500",
        conds.join(" AND ")
    );
    let mut offers: Vec<RawOffer> = Vec::new();
    {
        let mut rows = conn.query(sql.as_str(), ()).await.map_err(|e| format!("offers query: {e}"))?;
        while let Some(r) = rows.next().await.map_err(|e| format!("offer row: {e}"))? {
            let ppp: Option<i64> = r.get(2).ok();
            let Some(price) = ppp else { continue };
            let hotel_name: String = r.get(5).unwrap_or_default();
            let name: String = r.get(6).unwrap_or_default();
            offers.push(RawOffer {
                file: r.get(0).unwrap_or_default(),
                source_id: r.get(1).unwrap_or_default(),
                price_per_person: price,
                currency: { let c: String = r.get(3).unwrap_or_default(); if c.is_empty() { "TWD".to_string() } else { c } },
                airline: r.get(4).unwrap_or_default(),
                hotel: if !hotel_name.is_empty() { hotel_name } else { name },
            });
        }
    }
    // sort by price asc (stable) — matches TS scanOffers .sort
    offers.sort_by_key(|o| o.price_per_person);

    if offers.is_empty() {
        println!("No offers found for region \"{}\".", opts.region);
        std::process::exit(1);
    }

    let itinerary = parse_itinerary(opts.itinerary.as_deref());

    let mut results: Vec<TrueCostOffer> = Vec::new();
    for o in &offers {
        let area_type = if !o.hotel.is_empty() { detect_hotel_area(&o.hotel, &opts.region, &hotel_areas) } else { "unknown".to_string() };
        let hotel_hub = area_type_to_hub(&area_type, &opts.region, &regions);
        let baggage = calc_baggage_cost();
        let (transport_cost, transport_time) = calc_transport_cost(&hotel_hub, &itinerary, &opts.region, opts.pax, &regions, opts.jpy_rate);
        let surcharge = js_round((baggage + transport_cost) as f64 / opts.pax as f64);
        let true_total = o.price_per_person + surcharge;
        results.push(TrueCostOffer {
            file: o.file.clone(),
            source_name: o.source_id.clone(),
            price_per_person: o.price_per_person,
            currency: o.currency.clone(),
            airline: o.airline.clone(),
            hotel: o.hotel.clone(),
            hotel_area_type: area_type,
            baggage_cost: baggage,
            transport_cost,
            transport_time_min: transport_time,
            true_total,
            breakdown: format!("pkg:{} bag:{} xport:{}", o.price_per_person, baggage, transport_cost),
        });
    }
    results.sort_by_key(|r| r.true_total);

    print_results(&results, opts);
    Ok(())
}

struct RawOffer {
    file: String,
    source_id: String,
    price_per_person: i64,
    currency: String,
    airline: String,
    hotel: String,
}

struct TrueCostOffer {
    file: String,
    source_name: String,
    price_per_person: i64,
    currency: String,
    airline: String,
    hotel: String,
    hotel_area_type: String,
    baggage_cost: i64,
    transport_cost: i64,
    transport_time_min: i64,
    true_total: i64,
    breakdown: String,
}

fn parse_itinerary(spec: Option<&str>) -> Vec<(String, i64)> {
    let mut out = Vec::new();
    if let Some(s) = spec {
        for part in s.split(',') {
            let p = part.trim();
            #[allow(clippy::collapsible_if)]
            if let Some((dest, days)) = p.split_once(':') {
                let dest = dest.trim();
                if !dest.is_empty() {
                    if let Ok(d) = days.parse::<i64>() {
                        out.push((dest.to_string(), d));
                    }
                }
            }
        }
    }
    out
}

fn detect_hotel_area(hotel: &str, region: &str, hotel_areas: &BTreeMap<String, BTreeMap<String, Vec<String>>>) -> String {
    if let Some(areas) = hotel_areas.get(region) {
        for (area_type, keywords) in areas {
            for kw in keywords {
                if hotel.contains(kw.as_str()) {
                    return area_type.clone();
                }
            }
        }
    }
    "unknown".to_string()
}

fn area_type_to_hub(area_type: &str, region: &str, regions: &BTreeMap<String, RegionData>) -> String {
    let Some(rd) = regions.get(region) else { return String::new() };
    for (hub_id, hub_type, _) in &rd.hubs {
        if hub_type == area_type {
            return hub_id.clone();
        }
    }
    for (hub_id, hub_type, _) in &rd.hubs {
        if hub_type == "central" {
            return hub_id.clone();
        }
    }
    String::new()
}

fn calc_baggage_cost() -> i64 {
    // FIT packages (the only kind scanOffers emits) include baggage → 0.
    0
}

fn lookup_route<'a>(from: &str, to: &str, routes: &'a BTreeMap<String, TransportRoute>) -> Option<&'a TransportRoute> {
    routes.get(&format!("{from}-{to}")).or_else(|| routes.get(&format!("{to}-{from}")))
}

fn calc_transport_cost(hotel_hub: &str, itinerary: &[(String, i64)], region: &str, pax: i64, regions: &BTreeMap<String, RegionData>, jpy2twd: f64) -> (i64, i64) {
    let Some(rd) = regions.get(region) else { return (0, 0) };
    if hotel_hub.is_empty() {
        return (0, 0);
    }
    let mut total_jpy = 0_i64;
    let mut total_min = 0_i64;
    // airport transfer: 2 trips
    #[allow(clippy::collapsible_if)]
    if let Some((airport_hub, _, _)) = rd.hubs.iter().find(|(_, t, _)| t == "airport") {
        if let Some(rt) = lookup_route(airport_hub, hotel_hub, &rd.routes) {
            total_jpy += rt.cost_jpy * pax * 2;
            total_min += rt.time_min * 2;
        }
    }
    for (dest, days) in itinerary {
        if let Some(rt) = lookup_route(hotel_hub, dest, &rd.routes) {
            total_jpy += rt.cost_jpy * pax * 2 * days;
            total_min += rt.time_min * 2 * days;
        }
    }
    (js_round(total_jpy as f64 * jpy2twd), total_min)
}

fn print_results(results: &[TrueCostOffer], opts: &TrueCostArgs) {
    // best_time = Math.min over transport_time_min (a real number, addition works).
    let best_time = results.iter().map(|r| r.transport_time_min).min().unwrap_or(0);

    println!("\n╔════════════════════════════════════════════════════════════════════════════════════╗");
    println!("║  TRUE COST COMPARISON: {}║", pad_end(&opts.region.to_uppercase(), 57));
    println!("╚════════════════════════════════════════════════════════════════════════════════════╝");
    println!("  Pax: {} | Date: {} | JPY rate: {}", opts.pax, opts.date.as_deref().unwrap_or("(best)"), fmt_rate(opts.jpy_rate));
    if let Some(it) = &opts.itinerary {
        println!("  Itinerary: {it}");
    }
    println!();

    println!("┌──────────┬──────────┬─────────┬───────────┬────────────┬──────────┬──────────────────────┐");
    println!("│ Source   │ Package  │ Baggage │ Transport │ TRUE/person│ Time     │ Hotel                │");
    println!("├──────────┼──────────┼─────────┼───────────┼────────────┼──────────┼──────────────────────┤");
    for r in results {
        let src = pad_end(&truncate(&r.source_name, 8), 8);
        let pkg = pad_start(&locale(r.price_per_person), 8);
        let bag = if r.baggage_cost > 0 { pad_start(&locale(r.baggage_cost), 7) } else { "      0".to_string() };
        let xport = if r.transport_cost > 0 { pad_start(&locale(r.transport_cost), 9) } else { "        0".to_string() };
        let total = pad_start(&locale(r.true_total), 10);
        let time_diff = r.transport_time_min - best_time;
        let time_str = if time_diff == 0 { "optimal ".to_string() } else { pad_end(&format!("+{time_diff}min"), 8) };
        let hotel = pad_end(&truncate(if r.hotel.is_empty() { "-" } else { &r.hotel }, 20), 20);
        println!("│ {} │ {} │ {} │ {} │ {} │ {} │ {} │", src, pkg, bag, xport, total, time_str, hotel);
    }
    println!("└──────────┴──────────┴─────────┴───────────┴────────────┴──────────┴──────────────────────┘");

    if let Some(best) = results.first() {
        println!("\n💡 Best true value: {} at {} {}/person", best.source_name, best.currency, locale(best.true_total));
        println!("   {}", best.breakdown);
        if !best.hotel.is_empty() {
            println!("   Hotel: {} ({})", best.hotel, best.hotel_area_type);
        }
        if !best.airline.is_empty() {
            println!("   Airline: {}", best.airline);
        }
        // cheapest by package price (TS: [...results].sort((a,b)=>a.ppp-b.ppp)[0])
        let cheapest_pkg = results.iter().min_by_key(|r| r.price_per_person).unwrap();
        if cheapest_pkg.file != best.file {
            println!("\n⚠️  Cheapest package ({} {}) is NOT the best true value!", cheapest_pkg.source_name, locale(cheapest_pkg.price_per_person));
            println!("   Hidden costs: baggage {} + transport {}", locale(cheapest_pkg.baggage_cost), locale(cheapest_pkg.transport_cost));
        }
    }
    println!();
}

// Match JS number formatting of the rate (e.g. 0.22 → "0.22").
fn fmt_rate(r: f64) -> String {
    format!("{r}")
}
